File: 2015/07/07.clj
(require '[clojure.string :refer (split-lines split)])

(set! *warn-on-reflection* true)
(set! *unchecked-math* true)

(def input-file "07.txt")

(def gate-fn
  {"NOT" bit-not
   "OR" bit-or
   "AND" bit-and
   "RSHIFT" bit-shift-right
   "LSHIFT" bit-shift-left})

(defn parse-line [line]
  (let [[left right] (split line #" -> ")
        args (split left #" ")
        expr (case (count args)
               1 (list nil args)
               2 (list (gate-fn (first args)) (drop 1 args))
               3 (list (gate-fn (second args)) (list (first args) (last args))))]
    [right expr]))

(def solve
  (memoize
    (fn [gate-map register]
      (let [[gate args] (gate-map register)]
        (cond
          (number? args) args
          (re-matches #"\d+" register) (Integer/parseInt register)
          (nil? gate) (solve gate-map (first args))
          :else (apply gate (map #(solve gate-map %) args)))))))

(def gate-map
  (->> (slurp input-file)
       split-lines
       (map parse-line)
       (reduce conj {})))

(def silver (solve gate-map "a"))
(def gold
  (-> gate-map
      (assoc "b" (list nil silver))
      (solve "a")))

(println "Day 07\n======")
(println silver)
(println gold)

File: 2015/07/07.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

class Gate
  attr_accessor :op, :args

  def initialize(str)
    @op = case str
      when /.*NOT.*/    then ->(x) { ~x }
      when /.*OR.*/     then ->(a, b) { a | b }
      when /.*AND.*/    then ->(a, b) { a & b }
      when /.*RSHIFT.*/ then ->(a, b) { a >> b }
      when /.*LSHIFT.*/ then ->(a, b) { a << b }
      else                   ->(x) { x }
      end

    @args = str.gsub(/(NOT|OR|AND|LSHIFT|RSHIFT)/, "").strip.split(/\s+/)
  end
end

gates = Hash.new

File.open("07.txt").each_line do |line|
  left, right  = line.chomp.split " -> "
  gates[right] = Gate.new(left)
end

def solve(gates, set = nil)
  vals = Hash.new
  vals["b"] = set if !set.nil? # part 2

  until vals.member? "a"
    gates.each do |key, gate|
      next if vals.member? key
      next if !gate.args.all? { |x| !gates.member?(x) || vals.member?(x) }
      args = gate.args.map { |x| if gates.member? x then vals[x] else x.to_i end }
      vals[key] = gate.op.(*args)
    end
  end

  return vals["a"]
end

silver = solve(gates)
gold   = solve(gates, silver)

puts(
"Day 07
======
✮: #{silver}
★: #{gold}"
)
